src/main/kotlin/io/emeraldpay/dshackle/upstream/bitcoin/BitcoinBlockValidator.kt

```diff
@@ -0,0 +1,75 @@
+package io.emeraldpay.dshackle.upstream.bitcoin
+
+import io.emeraldpay.dshackle.data.BlockContainer
+import io.emeraldpay.dshackle.upstream.BlockValidator
+import io.emeraldpay.dshackle.upstream.ByteUtils.Companion.fromHexString
+import io.emeraldpay.dshackle.upstream.asUInt32
+import io.emeraldpay.dshackle.upstream.asUInt32LE
+import io.emeraldpay.dshackle.upstream.sha256
+import io.emeraldpay.dshackle.upstream.toHexString
+import org.apache.commons.codec.binary.Hex
+import org.slf4j.LoggerFactory
+import java.math.BigInteger
+
+class BitcoinBlockValidator : BlockValidator {
+    companion object {
+        private val log = LoggerFactory.getLogger(BitcoinBlockValidator::class.java)
+    }
+
+    override fun isValid(currentHead: BlockContainer?, newHead: BlockContainer): Boolean =
+        newHead.getParsed(Map::class.java)?.let { data ->
+            val bits = data["bits"] as String
+            val hash = data["hash"] as String
+            val hashValid = validateHash(bits, hash, data)
+            if (!hashValid) {
+                log.warn("Invalid header hash for block {}", hash)
+            }
+            val powValid = validatePoW(bits, hash)
+            if (!powValid) {
+                log.warn("Invalid PoW for block {}", hash)
+            }
+            hashValid && powValid
+        } ?: false
+
+    private fun validateHash(bits: String, hash: String, data: Map<*, *>): Boolean {
+        val version = data["version"] as Number
+        val prev = data["previousblockhash"] as String
+        val merkleRoot = data["merkleroot"] as String
+        val time = data["time"] as Number
+        val nonce = data["nonce"] as Number
+
+        val headerBytes = version.toLong().asUInt32()
+            .plus(fromHexString(prev).reversedArray())
+            .plus(fromHexString(merkleRoot).reversedArray())
+            .plus(time.toLong().asUInt32())
+            .plus(Hex.decodeHex(bits).asUInt32LE())
+            .plus(nonce.toLong().asUInt32())
+
+        return headerBytes.sha256().sha256().reversedArray().toHexString().lowercase() == hash
+    }
+
+    private fun validatePoW(bits: String, hash: String): Boolean {
+        val target = readCompactForm(bits.toLong(16))
+        val h = BigInteger(hash, 16)
```

Comment:
We also need some heuristic for chainwork validation, also we cannot validate it 100%, there must be someway to reduce the risk of spoofing this parameter

Reply:
added chainwork validation:

- chain work calculation check 
- chain work target change period check

Reply:
We can assume that bitcoin changes its target with fixed period, and if we know one day when it changed (and have constant for it), we can calculate the next target update